pkg/dataobj/consumer/logsobj/builder.go

```diff
@@ -223,10 +238,33 @@ func (b *Builder) Append(stream logproto.Stream) error {
 	}
 
 	b.currentSizeEstimate = b.estimatedSize()
-	b.state = builderStateDirty
+	if b.state == builderStateEmpty {
+		b.state = builderStateDirty
+		b.firstAppendedAt = now
```

Comment:
I specifically chose `now` instead of the record timestamp. If we used the record timestamp, when replaying a partition (i.e. following a crash), we would build one data object per record, which is not what we want.

Reply:
I think this might be better implemented at the consumer level, since that's also where we handle the idle timeout. WDYT?

Reply:
https://github.com/grafana/loki/pull/18622 exposes the time range for a builder which would make this possible, but nobody has been able to review it yet